### PR TITLE
Input Checking, Error Handling, and Twilio Input Validation

### PR DESCRIPTION
**Pull request by blujan**

### CODE REVIEW

**Comment by goldenapples** [comment]
The `validate_twilio_request` decorator works really well. 

I especially like the exception for `current_app.debug` so that we can still test out the api/message endpoint without having to go through the hassle of setting up an ngrok tunnel.  (I've been just testing message handling locally with httpie: `http post :5000/api/message From='+15035551234' Body='This is a test' -f`, and that works fine even after this change.) 

I'm wondering if there's a way we can add a similar exception to the `requires_auth` decorator. Since `requires_auth` also a current user, it might not be possible, but maybe we could have a separate level of authentication for GET requests, where we need to verify authentication, but don't necessarily need to know the current user.

I think I have a decent idea how to rework the client-side requests so that the user's auth token is sent along with every request. I can do that in a separate branch and open a PR into this one if you'd like...

File: application/app.py
Comment (goldenapples) [inline application/app.py]: At least right now, this can't require auth, because this signup is done before the user is logged in, and it returns the token used to auth for the rest of the session.

**Review by goldenapples — APPROVED** [review]
Looks good to me except for the create_user endpoint, which we have to work something else out for anyways. Are you going to do more here?

#22 should get the client sending the right authentication headers for all of these requests.

File: application/app.py
Comment (blujan) [reply application/app.py]: This wasn't final as I was also planning to do the front-end stuff (though, I see you got that!) and further tests, but yea, I can take that out for now.

**Comment by blujan** [comment]
Regarding the requires_auth decorator, I'm not quite sure what you mean. Was the goal of that so that we could do calls to it without authentication in debug mode or something else?

**Comment by goldenapples** [comment]
> Was the goal of that so that we could do calls to it without authentication in debug mode or something else?

Just that. I think this primarily applies to GET requests, where the current user isn't necessary - because most of the POST endpoints will error anyways if `g.current_user` isn't set. As we start passing more information in the API responses, or ordering/filtering results in the responses, I think it'd be nice to be able to check what the API is returning without having to run the front-end app.

**Comment by blujan** [comment]
Yea, I could see that being using for both manually testing during development and maybe later on if we want to write CI tests to validate the backend API. I'll take a look and see what might make sense.

**Comment by goldenapples** [comment]
@blujan Is this branch ready to merge? I cleared up the conflicts after merging #18.

**Pull request merged**
